DAP-02: messages

cjpatton opened the pull request:
Partially addresses #100.

Changes in this PR (each corresponds to a commit, so best to review each commit individually):
1. Bump daphne to 0.2.0 and  lock daphne_worker to daphne 0.1.2 until we're ready to upgrade to 0.2.0. This allows us to move daphne to DAP-02 without impacting daphne_worker.
2. Rework encoding of length-prefixed fields. In preparation for DAP-02, generalize the encoding methods to support     different prefix lengths.
3.  Align various messages with DAP-02: move the timestamp out of the nonce; define the ReportMetadata structure; update length prefixes; and add "batch-saturated" transition failure type. While at it, add type defs for Time and Duration and fiddle with buffers for AAD computations. (*)
4. Define BatchSelector and Query structs.
5. Add public share to Report and ReportShare. (*)
6. add report count to CollectResp. (*)

(*) Indicates a wire-breaking change.

cjpatton commented on daphne/src/lib.rs:
```diff
@@ -398,7 +401,7 @@ pub struct DapLeaderUncommitted {
 #[derive(Clone, Debug, PartialEq)]
 pub struct DapHelperState {
-    pub(crate) seq: Vec<(VdafState, Nonce)>,
+    pub(crate) seq: Vec<(VdafState, Time, Nonce)>,
```
This change is due to moving the `time` field out of the `Nonce` struct.

cjpatton commented on daphne/src/messages.rs:
```diff
@@ -4,7 +4,7 @@
 //! Messages in the DAP protocol.
```
Please double check that all of the following changes match the spec:
- Move time out of Nonce struct
- New ReportMetadata struct and corresponding changes to Report and ReportShare
- New field in AggregateInitialzeReq (we call it the BatchParameter here)
- All length prefixes are bumped properly (16 -> 32)
- New Query and BatchSelector structs (these are identical in fact).
- Changes to AggregateShareReq, CollectReq, and CollectResp.

cjpatton commented on daphne/src/messages.rs:
```diff
@@ -688,15 +825,84 @@ pub(crate) fn constant_time_eq(left: &[u8], right: &[u8]) -> bool {
     r == 0
 }
 
```
I'm planning to rework message serialization in the future: see #15.

chris-wood commented on daphne/src/messages.rs:
```diff
@@ -16,6 +16,8 @@ const KEM_ID_X25519_HKDF_SHA256: u16 = 0x0020;
 const KEM_ID_P256_HKDF_SHA256: u16 = 0x0010;
 const KDF_ID_HKDF_SHA256: u16 = 0x0001;
 const AEAD_ID_AES128GCM: u16 = 0x0001;
+const QUERY_TYPE_TIME_INTERVAL: u8 = 0x01;
+const QUERY_TYPE_FIXED_SIZE: u8 = 0x02;
```
I think these both should be `u16`s:

```
enum {
   reserved(0), /* Reserved for testing purposes */
   time_interval(1),
   fixed_size(2),
   (65535)
} QueryType;
```

cjpatton replied:
Oh god, good catch. Why did we make this 2 bytes instead of just one?

chris-wood commented on daphne/src/messages.rs:
```diff
@@ -389,36 +461,91 @@ impl Encode for Interval {
+        if let BatchSelector::TimeInterval { ref batch_interval } = self {
+            batch_interval
+        } else {
+            panic!("TOOD(issue #100)");
```
```suggestion
            panic!("TODO(issue #100)");
```
Panic with attitude!

chris-wood commented on daphne/src/messages.rs:
```diff
@@ -428,30 +555,40 @@ impl Decode for CollectReq {
 // TODO Add serialization tests.
 #[derive(Clone, Debug, Deserialize, PartialEq, Eq, Serialize)]
 pub struct CollectResp {
+    pub report_count: u64,
```
Isn't this missing the query-specific response parameters, like the batch ID for chunky queries?

cjpatton replied:
Yes, this gets added later in the PR stack with a type called PartialBatchSelector. Should be clearer then.

chris-wood replied:
Is that PR coming? I only see one PR in this stack.

cjpatton replied:
Yes.

cjpatton replied:
I haven't created PRs yet because there a number of them and I have to  keep them rebased.

chris-wood reviewed (changes requested):
Most of the DAP-02 message changes look good, though I think we missed a couple things.

cjpatton commented:
@chris-wood ready for another round. I updated the size of QueryType to match the spec. I also added a TODO to replace the message encoding stuff with what libprio will ship in 0.10.0.
